File: src/main/java/io/github/skippyall/minions/gui/minion/GuiContext.java
package io.github.skippyall.minions.gui.minion;

import io.github.skippyall.minions.minion.MinionRuntime;
import io.github.skippyall.minions.minion.fakeplayer.MinionFakePlayer;
import io.github.skippyall.minions.program.instruction.ConfiguredInstruction;
import io.github.skippyall.minions.program.supplier.Parameter;
import net.minecraft.server.network.ServerPlayerEntity;

public interface GuiContext {
    ServerPlayerEntity getViewer();

    static GuiContext create(ServerPlayerEntity viewer) {
        return new GuiContextImpl(viewer);
    }

    interface Minion extends GuiContext {
        MinionFakePlayer getMinion();

        static GuiContext.Minion create(GuiContext context, MinionFakePlayer minion) {
            return new GuiContextImpl.MinionImpl(context, minion);
        }
    }

    interface Instruction extends Minion {
        ConfiguredInstruction<MinionRuntime> getInstruction();

        String getName();

        void setName(String name);

        static GuiContext.Instruction create(GuiContext.Minion context, ConfiguredInstruction<MinionRuntime> instruction, String name) {
            return new GuiContextImpl.InstructionImpl(context, instruction, name);
        }
    }

    interface ValueSupplier extends Instruction {
        Parameter<?> getParameter();

        static GuiContext.ValueSupplier create(GuiContext.Instruction context, Parameter<?> parameter) {
            return new GuiContextImpl.ValueSupplierImpl(context, parameter);
        }
    }
}

File: src/main/java/io/github/skippyall/minions/gui/minion/GuiContext.kt
package io.github.skippyall.minions.gui.minion

import io.github.skippyall.minions.gui.minion.GuiContextImpl.InstructionImpl
import io.github.skippyall.minions.gui.minion.GuiContextImpl.MinionImpl
import io.github.skippyall.minions.gui.minion.GuiContextImpl.ValueSupplierImpl
import io.github.skippyall.minions.minion.MinionRuntime
import io.github.skippyall.minions.minion.fakeplayer.MinionFakePlayer
import io.github.skippyall.minions.program.instruction.ConfiguredInstruction
import io.github.skippyall.minions.program.supplier.Parameter
import net.minecraft.server.level.ServerPlayer

interface GuiContext {
    val viewer: ServerPlayer

    companion object {
        @JvmStatic
        fun create(viewer: ServerPlayer): GuiContext {
            return GuiContextImpl(viewer)
        }
    }

    interface Minion : GuiContext {
        val minion: MinionFakePlayer

        companion object {
            @JvmStatic
            fun create(context: GuiContext, minion: MinionFakePlayer): Minion {
                return MinionImpl(
                    if(context is MinionImpl) context.context else context,
                    minion
                )
            }
        }
    }

    interface Instruction : Minion {
        val instruction: ConfiguredInstruction

        var name: String

        companion object {
            @JvmStatic
            fun create(context: Minion, instruction: ConfiguredInstruction, name: String): Instruction {
                return InstructionImpl(
                    if(context is InstructionImpl) context.context else context,
                    instruction,
                    name
                )
            }
        }
    }

    interface ValueSupplier : Instruction {
        val parameter: Parameter<*>

        companion object {
            @JvmStatic
            fun create(context: Instruction, parameter: Parameter<*>): ValueSupplier {
                return ValueSupplierImpl(
                    if(context is ValueSupplierImpl) context.context else context,
                    parameter
                )
            }
        }
    }
}
